File: sentencex/languages/ta.py
from sentencex.base import Language

from .en import English

vowel_signs = {"ா", "ி", "ீ", "ு", "ூ", "ெ", "ே", "ை", "ொ", "ோ", "ௌ"}
vowels = {"அ", "ஆ", "இ", "ஈ", "உ", "ஊ", "எ", "ஏ", "ஐ", "ஒ", "ஓ", "ஔ"}
consonants = {
    "க",
    "ங",
    "ச",
    "ஞ",
    "ட",
    "ண",
    "த",
    "ந",
    "ப",
    "ம",
    "ய",
    "ர",
    "ல",
    "வ",
    "ழ",
    "ள",
    "ற",
    "ன",
}
consonant_vowels = set()

for consonant in consonants:
    for vowel_sign in vowel_signs:
        consonant_vowels.add(consonant + vowel_sign)


class Tamil(Language):
    language = "ta"

    # Writing English abbreviation like Dr as such inside Tamil is common
    abbreviations = (
        English.abbreviations.union(vowels)
        .union(consonants)
        .union(consonant_vowels)
        .union(
            {
                "ஏ",
                "பி",
                "சி",
                "டி",
                "ஈ",
                "எஃப்",
                "ஜி",
                "ஹேச்",
                "ஐ",
                "ஜே",
                "கே",
                "எல்",
                "எம்",
                "என்",
                "ஓ",
                # "பி",
                "க்யூ",
                "ஆர்",
                "எஸ்",
                # "டி",
                "யூ",
                "வி",
                "டபிள்யூ",
                "எக்ஸ்",
                "வை",
                "ஜெட்",
            }
        )
    )

File: src/languages/ta.rs
use std::sync::LazyLock;

use super::Language;

#[derive(Debug, Clone)]
pub struct Tamil {}
static TAMIL_ABBREVIATIONS: LazyLock<Vec<String>> = LazyLock::new(|| {
    let vowel_signs = vec!["ா", "ி", "ீ", "ু", "ূ", "ে", "ে", "ৈ", "ও", "ো", "ৌ"];
    let vowels = vec!["அ", "ஆ", "இ", "ஈ", "உ", "ஊ", "எ", "ஏ", "ஐ", "ஒ", "ஓ", "ஔ"];
    let consonants = vec![
        "க", "ங", "ச", "ஞ", "ட", "ண", "த", "ந", "ப", "ம", "ய", "ர", "ல", "வ", "ழ", "ள", "ற", "ன",
    ];

    let mut consonant_vowels = Vec::new();
    for consonant in &consonants {
        for vowel_sign in &vowel_signs {
            consonant_vowels.push(format!("{}{}", consonant, vowel_sign));
        }
    }

    include_str!("./abbrev/ta.txt")
        .lines()
        .chain(include_str!("./abbrev/en.txt").lines())
        .map(|line| line.trim().to_string())
        .filter(|line| !line.starts_with("//") && !line.is_empty())
        .collect::<Vec<String>>()
        .into_iter()
        .chain(vowels.into_iter().map(String::from))
        .chain(consonants.into_iter().map(String::from))
        .chain(consonant_vowels)
        .collect()
});

impl Language for Tamil {
    fn get_abbreviations(&self) -> &[String] {
        &TAMIL_ABBREVIATIONS
    }
}

#[cfg(test)]
mod tests {
    use crate::languages::tests::run_language_tests;

    use super::*;

    #[test]
    fn test_segment() {
        run_language_tests(Tamil {}, "tests/ta.txt");
    }
}
